### PR TITLE
[CUDA] Support memory reuse for dynamic shared memory

### PR DESCRIPTION
**Pull request by jinhongyii**
This PR applies part of the algorithm of storage rewrite to detect memory reuse possibility for dynamic shared memory, and does rewrite to the body. This functionality is important in matmul because we may want to allocate the shared memory for data fetch and the shared memory for write back in the same location.

### CODE REVIEW

**Comment by masahi** [comment]
Hmm, for reuse on the constant-size dynamic shmem, I thought the existing `storage_write` pass would do the job, https://github.com/apache/tvm/pull/8571#issuecomment-889541656. I wonder why it is not sufficient?

**Comment by vinx13** [comment]
I think this is to support reusing heterogenous data type allocs

**Comment by jinhongyii** [comment]
@masahi 
1. storage_rewrite can't handle storage scope with tag, and shared.dyn has tag
2. storage_rewrite can't handle buffer with different dtypes
3. storage_rewrite can't allocate a buffer in the place of another 2 buffers.  For example, you can take a look at https://github.com/jinhongyii/tvm/blob/dyn-shared-mem/tests/python/unittest/test_tir_transform_merge_dynamic_shared_memory_allocations.py#L80-L104
   C_sh's allocation reuses the memory of A_sh **and** B_sh. However, if we use the original algorithm of storage rewrite, it can only allocate C_sh in the place of A_sh **or** B_sh. The result allocation size would be block * block * 6

**Comment by junrushao1994** [comment]
Let's double check and get it in if there is no further issue :-)

File: src/tir/transforms/merge_dynamic_shared_memory_allocations.cc [hunk 1]
```diff
@@ -31,51 +31,256 @@
...
+// before_scope -> scope_body -> after_scope
+//
+// This pass tries to detect last point that we need to keep memory
+// alive under the same scope as Allocate.
+// The storage need to be kept alive between Allocate and last access.
+// The free point is only inserted at the same scope of Allocate.
+//
+class DynSharedMemLinearAccessPatternFinder final : public StmtExprVisitor {
+ public:
+  /*! \brief record the touch list of statment. */
```
Comment (vinx13) [inline src/tir/transforms/merge_dynamic_shared_memory_allocations.cc:78]:
```suggestion
  /*! \brief record the touch list of statement. */
```

**Review by vinx13 — COMMENTED** [review]
Otherwise Lgtm

**Review by vinx13 — APPROVED** [review]

**Review by Hzfengsy — APPROVED** [review]
Thanks @jinhongyii

**Pull request merged**